src/VisualStudio/Core/Def/Implementation/LanguageClient/VisualStudioLspSolutionProvider.cs

```diff
@@ -3,45 +3,53 @@
 using Microsoft.CodeAnalysis.LanguageServer;
-using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageClient
 {
     [Export(typeof(ILspSolutionProvider)), Shared]
     internal class VisualStudioLspSolutionProvider : ILspSolutionProvider
     {
-        private readonly VisualStudioWorkspace _visualStudioWorkspace;
-        private readonly MiscellaneousFilesWorkspace _miscellaneousFilesWorkspace;
+        private readonly ILspWorkspaceRegistrationService _registrationService;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public VisualStudioLspSolutionProvider(VisualStudioWorkspace visualStudioWorkspace, MiscellaneousFilesWorkspace miscellaneousFilesWorkspace)
+        public VisualStudioLspSolutionProvider(ILspWorkspaceRegistrationService registrationService)
         {
-            _visualStudioWorkspace = visualStudioWorkspace;
-            _miscellaneousFilesWorkspace = miscellaneousFilesWorkspace;
+            _registrationService = registrationService;
         }
 
-        public Solution GetCurrentSolutionForMainWorkspace()
+        public (DocumentId?, Solution) FindDocumentAndSolution(TextDocumentIdentifier? textDocument, string? clientName)
         {
-            return _visualStudioWorkspace.CurrentSolution;
-        }
+            // Assume the first workspace registered is the main one
+            var mainWorkspace = _registrationService.GetAllRegistrations().First();
 
-        public ImmutableArray<Document> GetDocuments(Uri documentUri)
-        {
-            // First check the VS workspace for matching documents.
-            var documents = _visualStudioWorkspace.CurrentSolution.GetDocuments(documentUri);
-            if (documents.IsEmpty)
+            // if we weren't asked for a document, then we just return the main solution
+            if (textDocument is null)
             {
-                // If there's none in the VS workspace, then check the misc files workspace.
-                documents = _miscellaneousFilesWorkspace.CurrentSolution.GetDocuments(documentUri);
+                return (null, mainWorkspace.CurrentSolution);
+            }
+
+            foreach (var workspace in _registrationService.GetAllRegistrations())
+            {
+                var documents = workspace.CurrentSolution.GetDocuments(textDocument.Uri, clientName);
+
+                if (!documents.IsEmpty)
+                {
+                    var document = documents.FindDocumentInProjectContext(textDocument);
+
+                    return (document.Id, document.Project.Solution);
+                }
             }
 
-            return documents;
+            // If we couldn't find the document then we don't know what solution, so just return the main one
+            return (null, mainWorkspace.CurrentSolution);
```

Comment:
The "null" return here does throw me off as well because it looked earlier that the API was generally a contract where you get non-null documents if you passed in a path, null document ids if you didn't get one. Again the two APIs really seems simpler.